File: scala/com/kotori316/fluidtank/tiles/Tiers.scala
package com.kotori316.fluidtank.tiles

import net.minecraft.nbt.NBTTagCompound

import scala.collection.JavaConverters._
import scala.collection.mutable

class Tiers private(val rank: Int, buckets: Int, override val toString: String, val meta: Int, val oreName: String) {
    val amount = buckets * 1000
    Tiers.map.put(toString, this)
    Tiers.list.append(this)
    Tiers.rankList(rank) = Tiers.rankList(rank) + 1

    override def hashCode(): Int = (rank << 20 + amount) ^ toString.hashCode

    def toNBTTag: NBTTagCompound = {
        val nbt = new NBTTagCompound
        nbt.setString("string", toString)
        nbt
    }
}

object Tiers {
    val rankList = Array.fill(9)(0)

    val map = mutable.Map.empty[String, Tiers]
    val list = mutable.ArrayBuffer.empty[Tiers]

    val Invalid = new Tiers(0, 0, "Invalid", 0, "Unknown")
    val WOOD = new Tiers(1, 1 << 2, "Wood", 0, "logWood")
    val STONE = new Tiers(2, 1 << 4, "Stone", 0, "stone")
    val IRON = new Tiers(3, 1 << 8, "Iron", 0, "ingotIron")
    val GOLD = new Tiers(4, 1 << 12, "Gold", 0, "ingotGold")
    val DIAMOND = new Tiers(5, 1 << 14, "Diamond", 0, "gemDiamond")
    val EMERALD = new Tiers(6, 1 << 16, "Emerald", 0, "gemEmerald")
    val STAR = new Tiers(7, 1 << 20, "Star", 0, "netherStar")
    val CREATIVE = new Tiers(8, 0, "Creative", 0, "Unknown") {
        override val amount: Int = Int.MaxValue
    }

    val COPPER = new Tiers(2, 1 << 5, "Copper", 1, "ingotCopper")
    val TIN = new Tiers(2, 1 << 6, "Tin", 2, "ingotTin")
    val BRONZE = new Tiers(3, 1 << 9, "Bronze", 1, "ingotBronze")
    val LEAD = new Tiers(3, 1 << 8, "Lead", 2, "ingotLead")
    val SILVER = new Tiers(3, 1 << 10, "Silver", 3, "ingotSilver")

    def jList: java.util.List[Tiers] = list.asJava

    def fromNBT(nbt: NBTTagCompound): Tiers = {
        val key = nbt.getString("string")
        map.getOrElse(key, {
            println("Invaluid pattern returned.")
            (new Exception).printStackTrace()
            WOOD
        })
    }
}

File: src/main/scala/com/kotori316/fluidtank/tank/Tiers.java
package com.kotori316.fluidtank.tank;

import java.util.Collections;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import net.fabricmc.fabric.api.tag.convention.v1.ConventionalItemTags;
import net.minecraft.core.Registry;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.tags.TagKey;
import net.minecraft.world.item.Item;
import org.apache.commons.lang3.tuple.Pair;
import org.jetbrains.annotations.Nullable;

import com.kotori316.fluidtank.FluidAmount;
import com.kotori316.fluidtank.TankConstant;

public enum Tiers {
    Invalid(0, 0, "Invalid", null),
    WOOD(1, 1 << 2, "Wood", null),
    STONE(2, 1 << 4, "Stone", MaterialTags.MATERIAL_STONE.location()),
    IRON(3, 1 << 8, "Iron", ConventionalItemTags.IRON_INGOTS.location()),
    GOLD(4, 1 << 12, "Gold", ConventionalItemTags.GOLD_INGOTS.location()),
    DIAMOND(5, 1 << 14, "Diamond", ConventionalItemTags.DIAMONDS.location()),
    EMERALD(6, 1 << 16, "Emerald", ConventionalItemTags.EMERALDS.location()),
    STAR(7, 1 << 20, "Star", MaterialTags.MATERIAL_STAR.location()),
    CREATIVE(8, 0, "Creative", null) {
        @Override
        public long amount() {
            return Long.MAX_VALUE;
        }
    },
    VOID(0, 0, "Void", null),
    COPPER(2, 1 << 5, "Copper", ConventionalItemTags.COPPER_INGOTS.location()),
    TIN(2, 1 << 6, "Tin", new ResourceLocation("c:tin_ingots")),
    BRONZE(3, 1 << 9, "Bronze", new ResourceLocation("c:bronze_ingots")),
    LEAD(3, 1 << 8, "Lead", new ResourceLocation("c:lead_ingots")),
    SILVER(3, 1 << 10, "Silver", new ResourceLocation("c:silver_ingots")),
    ;
    public static final Map<String, Tiers> TIERS_MAP;

    static {
        TIERS_MAP = Collections.unmodifiableMap(
            Stream.of(Tiers.values()).map(tiers -> Pair.of(tiers.toString(), tiers))
                .collect(Collectors.toMap(Pair::getKey, Pair::getValue))
        );
    }

    public final int rank;
    public final int buckets;
    @Nullable
    private final ResourceLocation tagName;
    private final String name;

    Tiers(int rank, int buckets, String name, @Nullable ResourceLocation tagName) {
        this.rank = rank;
        this.buckets = buckets;
        this.name = name;
        this.tagName = tagName;
    }

    @Override
    public String toString() {
        return name;
    }

    public CompoundTag toNBTTag() {
        CompoundTag nbt = new CompoundTag();
        nbt.putString("string", toString());
        return nbt;
    }

    public static Tiers fromNBT(CompoundTag tag) {
        if (tag == null) {
            return Invalid;
        }
        String name = tag.getString("string");
        return TIERS_MAP.getOrDefault(name, WOOD);
    }

    public static Stream<Tiers> stream() {
        return Stream.of(Tiers.values());
    }

    public long amount() {
        return TankConstant.config.capacity.get(name()).orElse(buckets) * FluidAmount.AMOUNT_BUCKET();
    }

    public TagKey<Item> getTag() {
        if (this.tagName == null)
            throw new IllegalStateException("Can't create tag for tier " + this);
        return TagKey.create(Registry.ITEM_REGISTRY, this.tagName);
    }

    public boolean hasTagRecipe() {
        return this.tagName != null;
    }
}
